homeassistant/components/text/__init__.py

```diff
@@ -0,0 +1,215 @@
+        raise HomeAssistantError(
+            f"Value {value} for {entity.name} is too long (maximum length {entity.max})"
+        )
+    if entity.pattern is not None and re.compile(entity.pattern).match(value) is None:
+        raise HomeAssistantError(
+            f"Value {value} for {entity.name} doesn't match pattern {entity.pattern}"
+        )
+    await entity.async_set_value(value)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    component: EntityComponent[TextEntity] = hass.data[DOMAIN]
+    return await component.async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    component: EntityComponent[TextEntity] = hass.data[DOMAIN]
+    return await component.async_unload_entry(entry)
+
+
+class TextMode(StrEnum):
+    """Modes for text entities."""
+
+    PASSWORD = "password"
+    TEXT = "text"
+
+
+@dataclass
+class TextEntityDescription(EntityDescription):
+    """A class that describes text entities."""
+
+    native_min: int | None = None
+    native_max: int | None = None
+    mode: TextMode | None = None
+    pattern: str | None = None
+
+
+class TextEntity(Entity):
+    """Representation of a Text entity."""
+
+    entity_description: TextEntityDescription
+    _attr_mode: TextMode = TextMode.TEXT
+    _attr_native_value: str | None
+    _attr_native_min: int = 0
+    _attr_native_max: int = MAX_LENGTH_STATE_STATE
+    _attr_pattern: str | None = None
+    _attr_state: None = None
+
+    @property
+    def capability_attributes(self) -> dict[str, Any]:
+        """Return capability attributes."""
+        return {
+            ATTR_MODE: self.mode,
+            ATTR_MIN: self.min,
+            ATTR_MAX: self.max,
+            ATTR_PATTERN: self.pattern,
+        }
+
+    @property
+    @final
+    def state(self) -> str | None:
+        """Return the entity state."""
+        if self.native_value is None:
+            return None
+        if len(self.native_value) < self.min:
+            raise ValueError(
+                f"Value {self.native_value} is too short (minimum length {self.min})"
+            )
+        if len(self.native_value) > self.max:
+            raise ValueError(
+                f"Value {self.native_value} is too long (maximum length {self.max})"
+            )
+        if (
+            self.pattern is not None
+            and re.compile(self.pattern).match(self.native_value) is None
+        ):
+            raise ValueError(f"Value does not match expected pattern {self.pattern}")
+        return self.native_value
+
+    @property
+    def mode(self) -> TextMode:
+        """Return the mode of the entity."""
+        if (
+            hasattr(self, "entity_description")
+            and self.entity_description.mode is not None
+        ):
+            return self.entity_description.mode
+        return self._attr_mode
```

Comment:
`attr` wins over entity descriptions.

```suggestion
        if hasattr(self, "_attr_mode"):
            return self._attr_mode
        if hasattr(self, "entity_description"):
            return self.entity_description.mode
        return TextMode.TEXT
```